### ISSUE TITLE
cbl-mariner-1.0-shim-x64-20210401

### ISSUE THREAD

**Issue by christopherco** [rhboot/shim-review#150]
Make sure you have provided the following information:

 - [x] link to your code branch cloned from rhboot/shim-review in the form user/repo@tag - https://github.com/microsoft/shim-review/tree/cbl-mariner-shim-x64-20210407
 - [x] completed README.md file with the necessary information
 - [x] shim.efi to be signed
 - [x] public portion of your certificate(s) embedded in shim (the file passed to VENDOR_CERT_FILE)
 - [x] binaries, for which hashes are added do vendor_db ( if you use vendor_db and have hashes allow-listed )
 - [x] any extra patches to shim via your own git tree or as files
 - [x] any extra patches to grub via your own git tree or as files
 - [x] build logs

###### What organization or people are asking to have this signed:
Microsoft Corporation

###### What product or service is this for:
CBL-Mariner

###### Please confirm this as the origin your shim.
Confirmed. We are building from the shim-15.4 tar file.

###### What's the justification that this really does need to be signed for the whole world to be able to boot it:
CBL-Mariner is an Linux distribution, built to run first party Azure cloud and edge workloads.

###### How do you manage and protect the keys used in your SHIM?
We use a hardware HSM

###### Do you use EV certificates as embedded certificates in the SHIM?
No

###### If you use new vendor_db functionality, are any hashes allow-listed, and if yes: for what binaries ?
No

###### Is kernel upstream commit 75b0cea7bf307f362057cc778efe89af4c615354 present in your kernel, if you boot chain includes a Linux kernel ?
Yes, these commits are present in our linux kernel. We are baselined on 5.10 LTS stable tree.
Source tree here: https://github.com/microsoft/CBL-Mariner-Linux-Kernel/tree/rolling-lts/mariner/5.10.21.1
- https://github.com/microsoft/CBL-Mariner-Linux-Kernel/commit/824d0b6225f3fa2992704478a8df520537cfcb56
- https://github.com/microsoft/CBL-Mariner-Linux-Kernel/commit/1957a85b0032a81e6482ca4aab883643b8dae06e

###### fixed ?
Yes. Our GRUB2 is baselined on 2.06-rc1 which has these fixes.

###### Please provide exact SBAT entries for all SBAT binaries you are booting or planning to boot directly through shim
SHIM SBAT entry example:
```
sbat,1,SBAT Version,sbat,1,https://github.com/rhboot/shim/blob/main/SBAT.md
shim,1,UEFI shim,shim,1,https://github.com/rhboot/shim
shim.mariner,1,Microsoft,shim,15.4-1.cm1,https://github.com/microsoft/CBL-Mariner
```

GRUB2 SBAT entry example:
```
sbat,1,SBAT Version,sbat,1,https://github.com/rhboot/shim/blob/main/SBAT.md
grub,1,Free Software Foundation,grub,2.06~rc1,https://www.gnu.org/software/grub/
grub.mariner,1,Microsoft,grub2,2.06~rc1-1.cm1,https://github.com/microsoft/CBL-Mariner
```

##### Were your old SHIM hashes provided to Microsoft ?
N/A. This is our first SHIM review request.

##### grub2 bootloaders can not be verified ?
We have a new signing key for grub2. We did not sign any grub2 builds affected by these CVEs with this new key. Our signed grub2 builds have the CVE fixes.

##### * Upstream grub2 shim_lock verifier or * Downstream RHEL/Fedora/Debian/Canonical like implementation ?
GRUB2 2.06-rc1 + Secure Boot patches from Fedora 34.

###### What is the origin and full version number of your bootloader (GRUB or other)?
Baseline is Grub 2.06-rc1 + Secure Boot patches from Fedora 34 + a few extra patches from Fedora 34

Our latest Grub2 spec and patches can be found here: https://github.com/microsoft/CBL-Mariner/tree/1.0-dev/SPECS/grub2

The current grub2 source RPM is also provided in this repository: [grub2-2.06~rc1-1.cm1.src.rpm](grub2-2.06~rc1-1.cm1.src.rpm)

###### If your SHIM launches any other components, please provide further details on what is launched
No other components launched by our shim.

###### please provide further details on what is launched and how it enforces Secureboot lockdown
N/A - we only launch the Linux kernel

###### apply. Please describe your strategy.
N/A. We changed to a new certificate.

###### How do the launched components prevent execution of unauthenticated code?
Grub2 has secure boot patches which will only load signed binaries.
All modules used are built into GRUB2 binary.

###### Does your SHIM load any loaders that support loading unsigned kernels (e.g. GRUB)?
No

###### What kernel are you using? Which patches does it includes to enforce Secure Boot?
We use the CBL-Mariner Linux Kernel, which is based on the 5.10 LTS kernel.
Current source tree here: https://github.com/microsoft/CBL-Mariner-Linux-Kernel/tree/rolling-lts/mariner/5.10.21.1

###### What changes were made since your SHIM was last signed?
N/A

###### What is the SHA256 hash of your final SHIM binary?
4930a1506341f7a37f4f22dd284d90e64bb4fb99ca8ee634b10ee52951e3604f  shimx64.efi

**Comment by steve-mcintyre** [comment]
Hi! Problem using your Dockerfile:

```
$ docker build .Sending build context to Docker daemon  219.4MB
Step 1/18 : FROM cbl-mariner-1.0:latest
pull access denied for cbl-mariner-1.0, repository does not exist or may require 'docker login'
```

**Comment by christopherco** [comment]
Hi @steve-mcintyre!
The `cbl-mariner-1.0:latest` base image is not published onto a public container registry at this time. To workaround this, in our submission repo, I added the `core-1.0.20210224.tar.gz` which is our minimal container base image, constructed from our packages published publicly.

Instructions on how to import this tar.gz into docker and tag the image can be found in the submission README [here](https://github.com/microsoft/shim-review/tree/cbl-mariner-shim-x64-20210401#what-os-and-toolchain-must-we-use-to-reproduce-this-build--include-where-to-find-it-etc--were-going-to-try-to-reproduce-your-build-as-close-as-possible-to-verify-that-its-really-a-build-of-the-source-tree-you-tell-us-it-is-so-these-need-to-be-fairly-thorough-at-the-very-least-include-the-specific-versions-of-gcc-binutils-and-gnu-efi-which-were-used-and-where-to-find-those-binariesif-possible-provide-a-dockerfile-that-rebuilds-the-shim)

**Comment by steve-mcintyre** [comment]
ah, ok. Thanks!

**Comment by jsetje** [comment]
The answers all look fine, and I am able to reproduce the shimx64.efi binary with the supplied docker image.

To account for the non-public docker image, I extracted the rpm version from the unpacked docker tar and the pulled the matching packages from the repos on packages.microsoft.com. This still shows some diffs, but they match what I would expect from different runs of the packaging system:

`Only in d/boot: systemd.cfg
Only in d/dev: null
diff -r c/etc/bash.bashrc d/etc/bash.bashrc
42d41
< export TERM=linux
Only in c/etc: machine-id
Binary files c/etc/pki/ca-trust/extracted/java/cacerts and d/etc/pki/ca-trust/extracted/java/cacerts differ
diff -r c/etc/shadow d/etc/shadow
1c1
< root:x:18725:0:::::
---
> root:x:18725:0:99999:7:::
12c12
< nobody:x:18725:0:99999:7:::
\ No newline at end of file
---
> nobody:x:18725:0:99999:7:::
Only in c/lib/debug: lib64
Only in c/lib/debug/usr: bin
Only in c/lib/debug/usr: lib
Only in c/lib/debug/usr: lib64
Only in c/lib/debug/usr: sbin
Only in c/lib64/debug: lib64
Only in c/lib64/debug/usr: bin
Only in c/lib64/debug/usr: lib
Only in c/lib64/debug/usr: lib64
Only in c/lib64/debug/usr: sbin
Only in c/root: .gnupg
Only in d/run: lock
Only in d/run: media
Only in c/tmp: postinstall
Only in c/usr/lib/debug: lib64
Only in c/usr/lib/debug/usr: bin
Only in c/usr/lib/debug/usr: lib
Only in c/usr/lib/debug/usr: lib64
Only in c/usr/lib/debug/usr: sbin
Only in c/usr/lib64/debug: lib64
Only in c/usr/lib64/debug/usr: bin
Only in c/usr/lib64/debug/usr: lib
Only in c/usr/lib64/debug/usr: lib64
Only in c/usr/lib64/debug/usr: sbin
Only in d/usr: src
Binary files c/var/cache/ldconfig/aux-cache and d/var/cache/ldconfig/aux-cache differ
Binary files c/var/lib/rpm/Basenames and d/var/lib/rpm/Basenames differ
Binary files c/var/lib/rpm/Conflictname and d/var/lib/rpm/Conflictname differ
Binary files c/var/lib/rpm/Dirnames and d/var/lib/rpm/Dirnames differ
Binary files c/var/lib/rpm/Enhancename and d/var/lib/rpm/Enhancename differ
Binary files c/var/lib/rpm/Filetriggername and d/var/lib/rpm/Filetriggername differ
Binary files c/var/lib/rpm/Group and d/var/lib/rpm/Group differ
Binary files c/var/lib/rpm/Installtid and d/var/lib/rpm/Installtid differ
Binary files c/var/lib/rpm/Name and d/var/lib/rpm/Name differ
Binary files c/var/lib/rpm/Obsoletename and d/var/lib/rpm/Obsoletename differ
Binary files c/var/lib/rpm/Packages and d/var/lib/rpm/Packages differ
Binary files c/var/lib/rpm/Providename and d/var/lib/rpm/Providename differ
Binary files c/var/lib/rpm/Recommendname and d/var/lib/rpm/Recommendname differ
Binary files c/var/lib/rpm/Requirename and d/var/lib/rpm/Requirename differ
Binary files c/var/lib/rpm/Sha1header and d/var/lib/rpm/Sha1header differ
Binary files c/var/lib/rpm/Sigmd5 and d/var/lib/rpm/Sigmd5 differ
Binary files c/var/lib/rpm/Suggestname and d/var/lib/rpm/Suggestname differ
Binary files c/var/lib/rpm/Supplementname and d/var/lib/rpm/Supplementname differ
Binary files c/var/lib/rpm/Transfiletriggername and d/var/lib/rpm/Transfiletriggername differ
Binary files c/var/lib/rpm/Triggername and d/var/lib/rpm/Triggername differ
Only in d/var/log: btmp
Only in d/var/log: journal
Only in c/var: mail
Only in c/var: mnt
Only in c/var: spool
Only in c/var: srv
`

**Comment by aburmash** [comment]
* build is reproducible. Produced binary's hash matches submission ( both hash in submission and hash of submitted binary )
Though i had to disable cgroupsv2 to actually run the container, for some weird reason my podman was not happy with it for your container image.

* SBAT section matches expectations
Contents of section .sbat:

> Contents of section .sbat:
 cb000 73626174 2c312c53 42415420 56657273  sbat,1,SBAT Vers
 cb010 696f6e2c 73626174 2c312c68 74747073  ion,sbat,1,https
 cb020 3a2f2f67 69746875 622e636f 6d2f7268  ://github.com/rh
 cb030 626f6f74 2f736869 6d2f626c 6f622f6d  boot/shim/blob/m
 cb040 61696e2f 53424154 2e6d640a 7368696d  ain/SBAT.md.shim
 cb050 2c312c55 45464920 7368696d 2c736869  ,1,UEFI shim,shi
 cb060 6d2c312c 68747470 733a2f2f 67697468  m,1,https://gith
 cb070 75622e63 6f6d2f72 68626f6f 742f7368  ub.com/rhboot/sh
 cb080 696d0a73 68696d2e 6d617269 6e65722c  im.shim.mariner,
 cb090 312c4d69 63726f73 6f66742c 7368696d  1,Microsoft,shim
 cb0a0 2c31352e 342d312e 636d312c 68747470  ,15.4-1.cm1,http
 cb0b0 733a2f2f 67697468 75622e63 6f6d2f6d  s://github.com/m
 cb0c0 6963726f 736f6674 2f43424c 2d4d6172  icrosoft/CBL-Mar
 cb0d0 696e6572 0a                          iner.  

* I had a look at grub2 repo provided and it looks good. Vanilla 2.06rc1 + downstream secureboot stack
@christopherco your submission implies that you used fc34 for reference, patches match and looking at how you handle sbat for grub my guess is that you had a look at redhat grub(s) ;) My only suggestion is to double check that nothing is missed, though i see all needed patches on place and grub module list looks modest, that's good, less stuff to worry about )


TLDR: looks good. Container was checked by @jsetje, so the only thing i did with container itself is download it and import to reproduce the build.

**Comment by julian-klode** [comment]
I'm happy with the reviews and marking accepted.

**Comment by julian-klode** [comment]
Closing issues labelled accepted
